### PR TITLE
ipq40xx: Add Mikrotik RouterBoard RB450Gx4 support 

### PR DESCRIPTION
**Pull request by robimarko**
RB450Gx4 is a 5 port gigabit router based on IPQ4019 + QCA8075.

Specification:

- Qualcomm IPQ4019 (717 MHz)
- 1024 MB of RAM (DDR3L)
- 2 MB (SPI NOR) + 512 MB (Parallel NAND) of flash
- 5x Gbit ethernet, POE IN on port 1 and POE passtrough to port 5.
- No WLAN radios enabled
- No USB ports
- micro SD card slot (Still requires work to get properly working)
- DB9 connector with UART on 115200 baud
- 2x LEDs
- GPIO Beeper
- 2x DC jack

Install instructions:
Since ipq-aux-loader enables native support its installed like ar71xx RB boards.
Only difference is that detach -p /dev/mtd1 and then ubiformat /dev/mtd1 are needed
before first sysupgrade.
After that normal sysupgrade will work.

Notes:
POE OUT is controlled via GPIO42, and needs to be manually exported, but be warned that due to essedma driver port that POE OUT is on is actually WAN port so be carefull.

Signed-off-by: Robert Marko <robimarko@gmail.com>

### CODE REVIEW

File: target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-rb450gx4.dts [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+	status = "okay";
+};
+
+&blsp1_spi1 {
+	pinctrl-0 = <&spi_0_pins>;
+	pinctrl-names = "default";
+	cs-gpios = <&tlmm 12 GPIO_ACTIVE_HIGH>;
+	status = "okay";
+
+	m25p80@0 {
```
Comment (dengqf6) [inline target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-rb450gx4.dts]: `flash@0`

File: target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-rb450gx4.dts [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			gpios = <&tlmm 5 GPIO_ACTIVE_HIGH>;
+			linux,code = <KEY_RESTART >;
```
Comment (dengqf6) [inline target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-rb450gx4.dts]: Trailing whitespace after `KEY_RESTART`

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-rb450gx4.dts [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			gpios = <&tlmm 5 GPIO_ACTIVE_HIGH>;
+			linux,code = <KEY_RESTART >;
```
Comment (dengqf6) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-rb450gx4.dts]: Ditto

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-rb450gx4.dts [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+	status = "okay";
+};
+
+&blsp1_spi1 {
+	pinctrl-0 = <&spi_0_pins>;
+	pinctrl-names = "default";
+	cs-gpios = <&tlmm 12 GPIO_ACTIVE_HIGH>;
+	status = "okay";
+
+	m25p80@0 {
```
Comment (dengqf6) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-rb450gx4.dts]: Ditto

**Comment by peterwillcn** [comment]
ping @dengqf6 @ynezz

**Comment by robimarko** [comment]
I gotta push fixes to DTS comments, but I don't know how to address your concerns about rbcfg as I am not the author, but rather have permission from the author to push those.
I will also drop 4.14 as ECC on NAND does not work under it and 4.19 is already default.

**Comment by dengqf6** [comment]
@adron-s
For the rbcfg part, it looks like the int32 is stored in little-endian byte order, and you tried to manually convert it to (or from) host byte order.
So why not use `le32toh` and `htole32` provided by `<endian.h>` instead? Just like `ntohs` and `htons`.

**Comment by zorun** [comment]
Hi @robimarko , thanks for your work!  This week-end, @hauke @mmaker and myself hacked on a hap-ac², and @hauke based its patches on your RB450Gx4 work. We managed to boot a working image on it.  I went back with the device so I will try to finish adding support for it in the coming weeks.

**Comment by robimarko** [comment]
@adrianschmutzler I will close this as I dont have time to rebase this.
I am hoping that LHGG and Habanero will at least get merged, but lately, ipq40xx has a lot of patches pending for months.

**Pull request closed without merging**